types/wlr_workspace_v1.c

```diff
@@ -0,0 +1,544 @@
+	group_output->group_handle = group;
+	wl_list_insert(&group->outputs, &group_output->link);
+
+	group_output->output_destroy.notify = toplevel_handle_output_destroy;
+	wl_signal_add(&output->events.destroy, &group_output->output_destroy);
+
+	group_send_output(group, output, true);
+}
+
+static void group_output_destroy(
+		struct wlr_workspace_group_handle_v1_output *output) {
+	wl_list_remove(&output->link);
+	wl_list_remove(&output->output_destroy.link);
+	free(output);
+}
+
+void wlr_workspace_group_handle_v1_output_leave(
+		struct wlr_workspace_group_handle_v1 *group, struct wlr_output *output) {
+	struct wlr_workspace_group_handle_v1_output *group_output_iterator;
+	struct wlr_workspace_group_handle_v1_output *group_output = NULL;
+
+	wl_list_for_each(group_output_iterator, &group->outputs, link) {
+		if (group_output_iterator->output == output) {
+			group_output = group_output_iterator;
+			break;
+		}
+	}
+
+	if (group_output) {
+		group_send_output(group, output, false);
+		group_output_destroy(group_output);
+	} else {
+		// XXX: log an error? crash?
+	}
+}
+
+static void group_send_details_to_resource(
+		struct wlr_workspace_group_handle_v1 *group,
+		struct wl_resource *resource) {
+	struct wlr_workspace_group_handle_v1_output *output;
+	wl_list_for_each(output, &group->outputs, link) {
+		send_output_to_group_resource(resource, output->output, true);
+	}
+}
+
+struct wlr_workspace_group_handle_v1 *wlr_workspace_group_handle_v1_create(
+		struct wlr_workspace_manager_v1 *manager) {
+
+	struct wlr_workspace_group_handle_v1 *group = calloc(1,
+			sizeof(struct wlr_workspace_group_handle_v1));
+	if (!group) {
+		return NULL;
+	}
+
+	group->manager = manager;
+	wl_list_insert(&manager->groups, &group->link);
+
+	wl_list_init(&group->outputs);
+	wl_list_init(&group->resources);
+	wl_list_init(&group->workspaces);
+	wl_signal_init(&group->events.destroy);
+
+	struct wl_resource *tmp, *manager_resource;
+	wl_resource_for_each_safe(manager_resource, tmp, &manager->resources) {
+		create_workspace_group_resource_for_resource(group, manager_resource);
+	}
+
+	return group;
+}
+
+void wlr_workspace_group_handle_v1_destroy(
+		struct wlr_workspace_group_handle_v1 *group) {
+	if (!group) {
+		return;
+	}
+
+	struct wlr_workspace_handle_v1 *workspace, *tmp;
+	wl_list_for_each_safe(workspace, tmp, &group->workspaces, link) {
+		wlr_workspace_handle_v1_destroy(workspace);
+	}
+
+	wlr_signal_emit_safe(&group->events.destroy, group);
+	workspace_manager_update_idle_source(group->manager);
+
+	struct wlr_workspace_group_handle_v1_output *output, *tmp2;
+	wl_list_for_each_safe(output, tmp2, &group->outputs, link) {
+		group_output_destroy(output);
+	}
+
+	struct wl_resource *tmp3, *resource;
+	wl_resource_for_each_safe(resource, tmp3, &group->resources) {
+		zwlr_workspace_group_handle_v1_send_remove(resource);
+
+		wl_resource_set_user_data(resource, NULL);
+		wl_list_remove(&resource->link);
+		wl_list_init(&resource->link);
```

Comment:
Why init here?

Reply:
Because the resource will be destroyed later here: https://github.com/swaywm/wlroots/pull/2357/files#diff-4adb740cfee4446f54fc395585e39a8fR247

We want that call there to be a valid list, otherwise, we may get a crash.

Reply:
I see.